### PR TITLE
SONAR-10145 Set built-in QG as default on existing organizations

### PR DESCRIPTION
**Pull request by ehartmann**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java [hunk 1]
```diff
@@ -49,10 +56,32 @@ QualityGateConditionDto getCondition(DbSession dbSession, long id) {
...
+      .setRequired(false)
+      .setInternal(false)
+      .setExampleValue("my-org");
+  }
+
+  /**
+   * @deprecated use {@link #getActions(OrganizationDto, QualityGateDto, QualityGateDto)} instead
+   */
+  @Deprecated
   Qualitygates.Actions getActions(QualityGateDto qualityGate, @Nullable QualityGateDto defaultQualityGate) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java [hunk 2]
```diff
@@ -49,10 +56,32 @@ QualityGateConditionDto getCondition(DbSession dbSession, long id) {
     return checkFound(dbClient.gateConditionDao().selectById(id, dbSession), "No quality gate condition with id '%d'", id);
   }
 
+  /**
+   * @deprecated use {@link #isQualityGateAdmin(OrganizationDto)} instead
+   */
+  @Deprecated
   boolean isQualityGateAdmin() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/CreateBuiltInQualityGate.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    this.uuidFactory = uuidFactory;
+  }
+
+
+  @Override
+  protected void execute(Context context) throws SQLException {
+    if (0 == context.prepareSelect("select count(uuid) from quality_gates where is_built_in = ?")
+      .setBoolean(1, true)
+      .get(row -> row.getLong(1))) {
+       context.prepareUpsert("insert into quality_gates (UUID, NAME, IS_BUILT_IN, CREATED_AT) values (?,?,?,?)")
```
Comment (SonarTech) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/CreateBuiltInQualityGate.java]: MAJOR Make this line start at column 7. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/PopulateDefaultQualityGateTest.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    insertOrganization(Uuids.createFast());
+
+    expectedException.expect(IllegalStateException.class);
+    expectedException.expectMessage("Unable to find the builtin quality gate");
+
+    underTest.execute();
+  }
+
+  @Test
+  public void should_not_fail_if_there_is_multiple_builtin_qualitygates() throws SQLException {
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/PopulateDefaultQualityGateTest.java]: If it would have been me, I would fail fast in this case => This is a strange situation where it's hard to predict what will happen

File: server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/PopulateOrgQualityGatesTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    insertOrganization(Uuids.createFast());
+
+    expectedException.expect(IllegalStateException.class);
+    expectedException.expectMessage("Unable to find the builtin quality gate");
+
+    underTest.execute();
+  }
+
+  @Test
+  public void should_not_fail_if_there_are_two_builtin_qg() throws SQLException {
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/PopulateOrgQualityGatesTest.java]: IMO, we should fail in this situation

File: server/sonar-server/src/main/java/org/sonar/server/organization/OrganizationCreationImpl.java [hunk 6]
```diff
@@ -186,7 +190,7 @@ private OrganizationDto insertOrganization(DbSession dbSession, NewOrganization
       .setKey(newOrganization.getKey())
       .setDescription(newOrganization.getDescription())
       .setUrl(newOrganization.getUrl())
-      // TODO .setDefaultQualityGateUuid("" + qualityGateFinder.getBuiltInQualityGate(dbSession).getId())
+      .setDefaultQualityGateUuid(qualityGateFinder.getBuiltInQualityGate(dbSession).getUuid())
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/organization/OrganizationCreationImpl.java]: This update should not have been done in another commit, ie "Set built-in QG as default of organization, where you have set the ORG#DEFAULT_QG column as nullable

Comment (julienlancelot) [reply server/sonar-server/src/main/java/org/sonar/server/organization/OrganizationCreationImpl.java]: Moreoever, it doesn't work, as you forgot to create an entry in ORG_QUALITY_GATES => No QGate are displayed on a new org

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/PopulateOrgQualityGates.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ */
+package org.sonar.server.platform.db.migration.version.v70;
+
+import java.sql.SQLException;
+import org.sonar.core.util.UuidFactory;
+import org.sonar.db.Database;
+import org.sonar.server.platform.db.migration.step.DataChange;
+import org.sonar.server.platform.db.migration.step.MassUpdate;
+
+public class PopulateOrgQualityGates extends DataChange {
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/PopulateOrgQualityGates.java]: The migration is missing quality gates created on default organization => After the migration, only the built-in QGate is seen on the default org

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/AssociateQualityGatesToDefaultOrganization.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    this.uuidFactory = uuidFactory;
+  }
+
+  @Override
+  protected void execute(Context context) throws SQLException {
+    String defaultOrgUuid = context.prepareSelect("select text_value from internal_properties where kee = 'organization.default'")
+      .get(row -> row.getString(1));
+
+    if (defaultOrgUuid == null) {
+      // No migration needed
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/AssociateQualityGatesToDefaultOrganization.java]: This should never happen => if you have a look in previous migrations, we fail in such situation (see for instance in https://github.com/SonarSource/sonarqube/blob/master/server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v62/PopulateOrganizationUuidOfGroupRoles.java#L58)

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/DbVersion70.java [hunk 9]
```diff
@@ -41,8 +41,11 @@ public void addSteps(MigrationStepRegistry registry) {
       .add(1911, "Add ORGANIZATIONS.DEFAULT_QUALITY_GATE_UUID", AddDefaultQualityGateUuidToOrganizations.class)
       .add(1912, "Create QUALITY_GATES.UUID", AddUuidToQualityGates.class)
       .add(1913, "Populate QUALITY_GATES.UUID", PopulateUuidOnQualityGates.class)
-      .add(1914, "Make QUALITY_GATES.UUID not nullable", MakeUuidNotNullableOnQualityGates.class)
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/DbVersion70.java]: Why have you drop this migration ?

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/DbVersion70.java [hunk 10]
```diff
@@ -41,8 +41,11 @@ public void addSteps(MigrationStepRegistry registry) {
...
-      .add(1915, "Drop unique index on QUALITY_GATES.NAME", DropUniqueIndexOnQualityGatesName.class)
+      .add(1914, "Drop unique index on QUALITY_GATES.NAME", DropUniqueIndexOnQualityGatesName.class)
+      .add(1915, "Create builtin quality gate if required", CreateBuiltInQualityGate.class)
+      .add(1916, "Populate ORG_QUALITY_GATES table", PopulateOrgQualityGates.class)
+      .add(1917, "Populate default quality gate on organization", PopulateDefaultQualityGate.class)
+      .add(1918, "Associate old quality gates to default organization", AssociateQualityGatesToDefaultOrganization.class)
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/DbVersion70.java]: Replace "old" by "existing"

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/PopulateDefaultQualityGate.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+  @Override
+  public void execute(Context context) throws SQLException {
+    String builtInQGUuid = context.prepareSelect("select uuid from quality_gates where is_built_in = ?")
+      .setBoolean(1, true)
+      .get(row -> row.getString(1));
+
+    checkState(builtInQGUuid != null, "Unable to find the builtin quality gate");
+
+    Long nbOfBuiltInQualityGates = context.prepareSelect("select count(uuid) from quality_gates where is_built_in = ?")
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/PopulateDefaultQualityGate.java]: This query is useless, you could have load all built-in quality gates on line 43 (replace .get() by .list())

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/PopulateOrgQualityGates.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      // No need for a migration
+      return;
+    }
+
+    String builtInQGUuid = context.prepareSelect("select uuid from quality_gates where is_built_in = ?")
+      .setBoolean(1, true)
+      .get(row -> row.getString(1));
+    checkState(builtInQGUuid != null, "Unable to find the builtin quality gate");
+
+    Long nbOfBuiltInQualityGates = context.prepareSelect("select count(uuid) from quality_gates where is_built_in = ?")
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/PopulateOrgQualityGates.java]: This query is useless and could be replaced by a .list() in previous query

File: server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/AssociateQualityGatesToDefaultOrganizationTest.java [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+  @Rule
+  public CoreDbTester db = CoreDbTester.createForSchema(AssociateQualityGatesToDefaultOrganizationTest.class, "org_quality_gates.sql");
+
+  private final UuidFactory uuidFactory = UuidFactoryFast.getInstance();
+
+  private AssociateQualityGatesToDefaultOrganization underTest = new AssociateQualityGatesToDefaultOrganization(db.database(), uuidFactory);
+
+  @Test
+  public void should_not_do_anything_if_no_default_organization() throws SQLException {
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/AssociateQualityGatesToDefaultOrganizationTest.java]: It should fail and not do anything :)

File: server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/AssociateQualityGatesToDefaultOrganizationTest.java [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    insertQualityGate(uuidFactory.create(), "QualityGate2", true);
+
+    underTest.execute();
+
+    assertThat(selectAllOrgQualityGates()).isEmpty();
+  }
+
+  @Test
+  public void should_associate_all_quality_gates_to_default_organization() throws SQLException {
+    String orgUuid1 = uuidFactory.create();
```
Comment (julienlancelot) [inline server/sonar-db-migration/src/test/java/org/sonar/server/platform/db/migration/version/v70/AssociateQualityGatesToDefaultOrganizationTest.java]: Rename to defaultOrgUuid => improve readability

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/DbVersion70.java (hunk 9 above)
Comment (julienlancelot) [reply server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/DbVersion70.java]: Moreoever, it's no more consistent with H2 => https://github.com/SonarSource/sonarqube/blob/feature/eh/SONAR-10145/server/sonar-db-core/src/main/resources/org/sonar/db/version/schema-h2.ddl#L232
